### PR TITLE
Quasi-static and dynamic Timoshenko Beam #10313

### LINKED ISSUE #10313: Euler-Bernoulli and Timoshenko beam element
## Rationale
Beam elements are useful in modeling long slender objects that are more susceptible to bending behavior. 

## Description
Added small strain, large strain and large rotation C0 timoshenko beam element (Euler-Bernoulli beam is a special case of timoshenko beam element)

## Impact
This does not change anything in the already existing code.

### PR DESCRIPTION
**Pull request by sveerara**
Strain and stress calculators and inertial force calculators have been added for the C0 Timoshenko beam element. 

Quasi-static and dynamic beam bending tests have been added. 

Nodal inertial force/torque calculators have been added and corresponding tests have been. 

closes #10313

### CODE REVIEW

**Comment by sveerara** [comment]
@bwspenc, all the tests are working fine now. Can you please review this PR when you have some time? Thanks!

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: Add 'The' in front of the last sentence.

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: Capitalize 'Timoshenko'.  Also, I think it's better to capitalize 'dof', but that's up to you.

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: strain->strains

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: remove 'onto'

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: I think I'd reword the last part of this:
'the second method assumes that the mass of the beam is concentrated at the ends of the beam, and represents the mass of the beam using point mass/inertia at the nodes at the end of the beam.'

File: docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md
Comment (bwspenc) [inline docs/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md]: Add 'the' after 'If'.

Also, the last part of the last sentence doesn't really make sense (the 'which' seems extraneous).

File: docs/content/documentation/systems/Kernels/tensor_mechanics/InertialForceBeam.md
Comment (bwspenc) [inline docs/content/documentation/systems/Kernels/tensor_mechanics/InertialForceBeam.md]: These files all need some descriptions of what the classes do.  These can be really short, and should all reference the very thorough documentation you provided in C0TimoshenkoBeam.md.

**Review by bwspenc — CHANGES_REQUESTED** [review]
This is as far as I've gotten so far. I'll keep working on this -- just wanted to give you feedback so you could fix a few minor things with the documentation while I'm going through the rest of it.  This is looking really good!

File: modules/tensor_mechanics/include/materials/ComputeElasticityBeam.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/ComputeElasticityBeam.h]: Since 'elasticity vector' isn't a standard thing, maybe better to explain that a little more.  I'd say something like:

ComputeElasticityBeam computes the equivalent of the elasticity tensor for the beam element, which are vectors of material translational and flexural stiffness.

File: modules/tensor_mechanics/include/nodalkernels/NodalInertialForce.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/nodalkernels/NodalInertialForce.h]: Rename this `NodalTranslationalInertia`?

File: modules/tensor_mechanics/include/nodalkernels/NodalInertialTorque.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/nodalkernels/NodalInertialTorque.h]: Rename `NodalRotationalInertia`?

File: modules/tensor_mechanics/include/nodalkernels/NodalInertialTorque.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/nodalkernels/NodalInertialTorque.h]: This description is for the translational version.

**Comment by sveerara** [comment]
@bwspenc I have added the eigenstrain classes and the eigenstrain transfer class. Please review the PR when you have time. Some test in Rattlesnake is failing. My commits don't change any existing classes so its probably not related to this PR.

File: modules/tensor_mechanics/include/transfers/VectorPostprocessorToAuxVar.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/transfers/VectorPostprocessorToAuxVar.h]: I'm looking at the other transfers to see if the naming convention for this matches those, and it seems that something like 'VectorPostprocessorTransfer' might be more consistent. None of the transfers have anything about what kind of thing they are transferring to in their name.

File: modules/tensor_mechanics/include/userobjects/VectorPostprocessorToInterpolator.h
Comment (bwspenc) [inline modules/tensor_mechanics/include/userobjects/VectorPostprocessorToInterpolator.h]: I think I would rename this to something like 'CSVInterpolator' or 'InterpolateFromCSV'. The CSV files happen to be generated by a VectorPostprocessor in your case, but they could really come from anywhere.

File: docs/content/documentation/systems/Materials/tensor_mechanics/ComputeFiniteBeamStrain.md
Comment (bwspenc) [inline docs/content/documentation/systems/Materials/tensor_mechanics/ComputeFiniteBeamStrain.md]: This description (and the one in the header for this class) makes it sound like it only computes the rotation matrix. It also computes the strain increments (inherited from the base class). Change the description to make that clear.

File: modules/combined/test/tests/beam_eigenstrain_transfer/master.i
Comment (bwspenc) [inline modules/combined/test/tests/beam_eigenstrain_transfer/master.i]: We should really add an Action like we have for the other StressDivergence kernels so you don't need to set up all of these (especially since we have 6 rather than 3 components).

It looks like we don't have that for the Truss ones either. Come to think of it, can we get the truss formulation by using a subset of the beam code?  It would be good to consolidate the truss and beam capabilities as much as possible (as long as it doesn't make the truss input more complex).

**Review by bwspenc — CHANGES_REQUESTED** [review]

**Comment by bwspenc** [comment]
@sveerara It sounds like the documentation system will be down until March 22. We can either wait until then to merge this, or strip out the Markdown files from this PR. Apparently there will be some changes to the Markdown file format (which is why it is frozen), but I imagine that it will be pretty easy to convert your stuff over.  Since it's a little ways off, I'd be inclined to go ahead and merge it before then without the Markdown.

File: modules/tensor_mechanics/src/materials/ComputeElasticityBeam.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ComputeElasticityBeam.C]: I'd prefer "scale factor" over "reduction factor" in this description.  Also, I think I'd change the descriptions of those coupled variables to say something like "Young's modulus of the material. Can be supplied as either a number or a variable name." ( so it's less confusing to the typical user who will probably just supply a number rather than a variable).

File: modules/tensor_mechanics/src/materials/ComputeBeamEigenstrainFromAuxVar.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ComputeBeamEigenstrainFromAuxVar.C]: I think I'd re-name this "ComputeBeamEigenstrainFromVariable" because it could work with either an AuxVariable or a Variable. Also, change the parameter descriptions to reflect that fact as well (remove 'aux').

**Comment by sveerara** [comment]
@bwspenc I will remove the documentation files when I address your comments. I was thinking of creating an action that creates  StressDivergenceBeam objects and also creates the NewmarkVelAux, NewmarkAccelAux and InertialForceBeam if `dynamic_consistent_mass` is set to true. And then maybe add a option for the nodal forces/inertia.

I will look at the truss files and see how it can be consolidated with the beam files. The beam can perform the functions of a truss as well but it would be storing more material properties than what a truss requires. If memory is not a problem, then I can make a truss class that inherits from the beam classes and set the unwanted strains/forces/moments to zero. I can also create separate classes for the truss using the plug-and-play model (strain calculator, elasticity and force calculator). In seperate classes are created, only axial force and strain need to be stored so it might be more memory efficient. Please let me know which method is better.

File: modules/tensor_mechanics/src/materials/ComputeBeamEigenstrainFromVPP.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ComputeBeamEigenstrainFromVPP.C]: Since it's not coming straight from a VPP, I think a name like "ComputeBeamEigenstrainFromCSVInterpolator" would be more clear (assuming you rename VectorPostprocessorToInterpolator to CSVInterpolator).

**Comment by bwspenc** [comment]
@sveerara I wonder if we could keep the current truss material model (since it's simpler and doesn't need the memory for as many properties), but make it so you can use a common Kernel for both the translational part of the beam formulation and the truss.

**Comment by sveerara** [comment]
@bwspenc and @acasagran, I have updated the PR.

**Comment by sveerara** [comment]
I will remove the documentation files and squash all the commits into one once the reviews are completed and the PR is ready to be merged.

File: modules/tensor_mechanics/test/tests/beam/action/tests
Comment (acasagran) [inline modules/tensor_mechanics/test/tests/beam/action/tests]: This should be type = 'Exodiff' both here and below.

File: modules/tensor_mechanics/test/tests/beam/action/2_block_common.i
Comment (acasagran) [inline modules/tensor_mechanics/test/tests/beam/action/2_block_common.i]: This should be '2_block_out' if you want to compare to the previous test which it looks like from the tests file.

**Comment by sveerara** [comment]
This PR has too many files so I am breaking it into smaller PRs.

**Pull request closed without merging**
